docs/jep/11859-type-annotations.md

```diff
@@ -0,0 +1,189 @@
+
+```python
+Array = Any
+Shape = core.Shape
+
+def slice(operand: Array, start_indices: Sequence[int],
+          limit_indices: Sequence[int],
+          strides: Optional[Sequence[int]] = None) -> Array:
+  ...
+```
+
+For the purposes of static type checking, this use of `Array = Any` for array annotations puts no constraint on the argument values (`Any` is equivalent to no annotation at all) but it does serve as a form of useful in-code documentation for the developer.
+
+For the sake of generated documentation, the name of the alias gets lost (the [HTML docs](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.slice.html) for `jax.lax.slice` report operand as type `Any`), so the documentation benefit does not go beyond the source code.
+
+A benefit of this level of type annotation is that it is never wrong to annotate a value with `Any`, so it will provide a concrete benefit to developers and users in the form of documentation, without added complexity of satisfying the stricter needs of any particular static type checker.
+
+### Level 2: annotations for intelligent autocomplete
+Many modern IDEs take advantage of type annotations to improve [intelligent code completion](https://en.wikipedia.org/wiki/Intelligent_code_completion) in IDEs. One example of this is the [PyLance](https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-pylance) extension for VSCode, which uses Microsofts' [pyright](https://github.com/microsoft/pyright) static type checker as a source of information for VSCode's [IntelliSense](https://code.visualstudio.com/docs/editor/intellisense) completions.
+
+This use of type checking requires going further than the simple aliases used above; for example, knowing that the `slice` function returns an alias of `Any` named `Array` does not add any useful information to the code completion engine. However, were we to annotate the function with a `DeviceArray` return type, the autocomplete would know how to populate the namespace of the result, and thus be able to suggest more relevant autocompletions during the course of development.
+
+JAX has begun to add this level of type annotation in a few places; one example is the `jnp.ndarray` return type within the `jax.random` package [[source](http://google3/third_party/py/jax/_src/random.py;l=359;rcl=445522278)]:
+
+```python
+def shuffle(key: KeyArray, x: Array, axis: int = 0) -> jnp.ndarray:
+  ...
+```
+
+In this case `jnp.ndarray` is an abstract base class that forward-declares the attributes and methods of JAX arrays ([see source](http://google3/third_party/py/jax/_src/numpy/ndarray.py;l=40;rcl=431999528)), and so PyLance in VSCode can offer the full set of autocompletions on results from this function. Here is the result in VSCode:
```

Comment:
As before. Maybe this will work better https://github.com/google/jax/blob/main/jax/_src/numpy/ndarray.py#L41